### ISSUE TITLE
review of package

### ISSUE THREAD

**Issue by theorashid** [juanitorduz/numpyro_forecast#75]
# numpyro_forecast: API feedback

I've gone through this at a high level through the docs (I trust you to write great numpyro), so I'm mostly asking "why, as an end user, would I not use this API?". The docs are great, obviously.

A lot of the below is personal preference (I prefer functional to OOP, as I've mentioned before). I know this is a port of the Pyro subpackage, so a lot of it goes against their design. Interestingly, a former colleague of mine worked on [Uber's Orbit](https://github.com/uber/orbit), which used the exact same `Forecaster` pattern as the Pyro package (I think some devs worked across both), and then brought the pattern to our team. I did not like it much and pushed away from it. So how much of this you take depends on how faithful you want to stay to Pyro.

## 1. API stuff: library, not framework

Writing numpyro normally feels lightweight, like using `jax.numpy`. This package feels heavier. I think because the examples have the whole framework (`Forecaster`, `fit_svi`, `draw_posterior`, `forecast`) rather than just using base numpyro.

I want a library of useful-for-time-series building blocks that work inside my model and under standard numpyro inference.

[dynestyx](https://github.com/BasisResearch/dynestyx) does this well. The user writes a plain numpyro model, calls `dsx.sample("f", dynamics, obs_times=..., obs_values=...)` inside it (same shape as `numpyro.sample`), then runs standard `MCMC(NUTS(model))` or `Predictive(model)`, with behaviour dictated by specific handlers (`with Filter():`).

The value from this package is `time_series`/`markov_time_series` (the `_future`-site trick that stops the guide resizing), `predict_glm` (prefix conditioning, including the proper MVN Gaussian conditional), which are already functions of this shape (they expand into `sample`/`plate` statements, same status as `numpyro.contrib.control_flow.scan`). The layers around that core (the forecaster classes, inference wrappers, model wrapper) make it feel like a whole new API.

### Why is there a `ForecastingModel` *and* a `Forecaster`? And why a `Forecaster` per inference method?

This is the part I didn't like from Orbit.

```py
forecaster = Forecaster(key_fit, SeasonalForecaster(), data, covariates[:t_obs], num_steps=1_500)
```

`ForecastingModel` already *is* the numpyro model (it's callable as `(covariates, data=None)`). `Forecaster` then bundles model + data + an SVI run + prediction into one stateful object whose constructor does the fitting. imo the model shouldn't need a class, and prediction shouldn't need an object that owns the data.

I also don't see the need for separate `Forecaster` / `HMCForecaster` / `PathfinderForecaster` because inference should be decoupled from the model.

### What do `forecasting_model` and `time_series` do?

Looking into the code, `forecasting_model` is two lines that computes `Horizon.from_data(covariates, data)` and passes it to your body.

Then parts which control training vs forecast: `time_series` (in-sample latent under a fixed site, forecast suffix under a separate `_future` site, so `AutoNormal` never resizes and `Predictive` draws the suffix from the prior) and `predict` (observe the prefix, sample the horizon via `prefix_condition`).

Some notes:

- `forecasting_model` is optional. Nothing stops me writing a plain model with `h = Horizon.from_data(covariates, data)` as the first line (which I prefer).
- `predict` is one line over `predict_glm` (a location-shift link), with an unenforced precondition: the noise must be zero-centered, and `predict(dist.Normal(3.0, sigma), ...)` silently biases the model. Why both?
- I'm not a fan of the name `time_series`. It samples conditionally *iid innovations* under a time plate, vs `markov_time_series` where the latent state is carried. I'd name them something like `innovations` and `markov`.

### proposal: some small numpyro_forecast functions inside a standard numpyro model

Something like:

```py
import numpyro_forecast as nf

def model(covariates, data=None):
    h = nf.horizon(covariates, data)  # rename Horizon.from_data; the one convention to learn
    # still not 100% sure on this but it does some useful shape checking

    drift_scale = numpyro.sample("drift_scale", dist.LogNormal(-3.0, 1.0))
    sigma = numpyro.sample("sigma", dist.LogNormal(-2.0, 1.0))

    drift = nf.innovations("drift", lambda: dist.Normal(0.0, drift_scale), h)
    level = jnp.cumsum(drift, axis=-2)

    nf.predict(lambda mu: dist.Normal(mu, sigma), level, obs=data)  # one predict, not predict_glm too
    # this also keeps obs=, which I like because that's where the data connection is
    # It's also similar to dynestyx
    # obs=None` is the prior predictive
    # full-length `obs` is training
    # shorter `obs` means forecast

```

Then, no need for `Forecaster` (nor `HMCForecaster`) nor `fit_svi`:

```py
guide = AutoNormal(model)
svi_result = SVI(model, guide, Adam(0.01), Trace_ELBO()).run(key_fit, 1_500, covariates[:t_obs], data)
posterior = guide.sample_posterior(key_post, svi_result.params, sample_shape=(100,))
samples = Predictive(model, posterior)(key_pred, covariates, data)["forecast"]
```

I think it's worth explaining these core parts (`Horizon.from_data`, `time_series`, `markov_time_series`, `predict_glm`, `register_elementwise`, or however they end up) and the numpyro that they replace in the docs as the core of the package. Everything else in the model layer is wrapping.

e.g. `predict` replaces about six lines of numpyro:

```py
T = data.shape[-2]
if prediction.shape[-2] == T:                      # training
    numpyro.sample("obs", dist.StudentT(nu, prediction, sigma), obs=data)
else:                                              # forecasting
    numpyro.sample("obs", dist.StudentT(nu, prediction[..., :T, :], sigma), obs=data)
    fc = numpyro.sample("obs_future", dist.StudentT(nu, prediction[..., T:, :], sigma))
    numpyro.deterministic("forecast", fc)
```

This also shows the users how these create sites that are never explicitly typed (`drift_future`, `obs_future`, `forecast`)

Aside: why don't we do this with `numpyro.handlers.scope` instead? Probably a more robust way of dealing with the namespace to get `future/drift`, `future/obs`.

## 2. Three time-series structures (and the missing one)

Got uncle Claude to compare the 12 example notebooks by which model-building functions they use:

| Notebooks | What they use |
|---|---|
| forecasting_univariate, hierarchical_1, hierarchical_2, inference_methods, electricity (x2) | `time_series` + `predict` |
| fresh_retail_stockout | `time_series` + `markov_time_series` + `predict_glm` |
| arma, exponential_smoothing, croston, tsb, availability_tsb | raw numpyro inside the wrapper (hand-rolled scans, hand-rolled `eps_future` sites, hand-registered `"forecast"`) |

All error-feedback models (ARMA, ETS in innovations form, Croston/TSB) bypass the model-building functions entirely. The three structures:

1. **iid innovations** (`time_series` today): conditionally iid latent, series built arithmetically. Prior-drawn forecast suffix is valid because of exchangeability.
2. **Markov latent state** (`markov_time_series` today): latent sampled per step, forecast scan seeded by the final in-sample carry.
3. **Error feedback / SSOE, the missing one**: one innovation per step is both the observation noise and the driver of the state update (`error = y_t - pred_t` in-sample; sampled and fed back at forecast time). `predict` can't do it (the horizon predictor depends on sampled innovations, so there's no deterministic full-horizon prediction) and `markov_time_series` can't either (nothing is sampled in-sample; the filter is deterministic given data). Hence five notebooks hand-rolling the same pattern.

One function would cover them:

```py
nf.ssoe(
    "state",
    init_carry=(level_init, trend_init, seasonality_init),
    predict_fn=lambda carry: ...,   # carry -> one-step-ahead prediction
    advance=advance,                # (carry, innovation) -> next carry
    noise=dist.Normal(0, sigma),
    obs=data,
    h=h,
)
```

The package owns the in-sample filter (innovation = residual, observe against data), the forecast scan (sample innovation, feed it back), and the site conventions; the user supplies the state mechanics in the component form forecasters think in. ARMA(1,1) is `predict_fn = lambda c: mu + phi*c[0] + theta*c[1]`, `advance = lambda c, e: (predict_fn(c) + e, e)`; ETS is the level/trend/seasonality update the notebook already has.

Which to use basically depends on whether `numpyro.sample` needs to live inside the recursion:

- innovations: one vectorized plate, outside any loop; recursion is none or deterministic (`cumsum`, `jax.lax.scan`) after sampling
- ssoe: one vectorized plate (iid innovations), outside the loop; deterministic `jax.lax.scan` consumes them via the carry
- markov: sampling per step, inside the loop; `numpyro.contrib.control_flow.scan` needed

Also, for the linear-Gaussian members (ARMA, additive ETS) we can marginalise out using a Kalman filter through dynestyx `with Filter():`. The SSOE recursion is probably the better form for the nonlinear members (multiplicative ETS, GARCH-type recursions, Croston/TSB on counts), which is what the notebooks hand-roll.

## worked example: the univariate notebook in the proposed style

The `forecasting_univariate` model is currently written twice in the notebook (once as a `ForecastingModel` subclass, once as a functional body wrapped in `forecasting_model`), which itself makes the point about two APIs. In the proposed style it's one plain numpyro model:

```py
import numpyro_forecast as nf

def univariate(covariates, data=None):
    h = nf.horizon(covariates, data)
    num_features = covariates.shape[-1]

    bias = numpyro.sample("bias", dist.Normal(0.0, 10.0))
    weight = numpyro.sample("weight", dist.Normal(0.0, 0.1).expand([num_features]).to_event(1))
    drift_scale = numpyro.sample("drift_scale", dist.LogNormal(-20.0, 5.0))
    nu = numpyro.sample("nu", dist.Gamma(10.0, 2.0))
    sigma = numpyro.sample("sigma", dist.LogNormal(-5.0, 5.0))
    centered = numpyro.sample("centered", dist.Uniform(0.0, 1.0))

    drift = nf.innovations(
        "drift",
        lambda: dist.Normal(0.0, drift_scale),
        h,
        reparam=LocScaleReparam(centered=centered),
    )
    level = jnp.cumsum(drift, axis=-2)
    regression = (weight * covariates).sum(axis=-1, keepdims=True)

    nf.predict(lambda mu: dist.StudentT(nu, mu, sigma), level + bias + regression, obs=data)
```

fitted and forecast with standard numpyro:

```py
mcmc = MCMC(NUTS(univariate), num_warmup=1_000, num_samples=1_000)
mcmc.run(key_fit, covariates[:t_obs], data)
samples = Predictive(univariate, mcmc.get_samples())(key_pred, covariates, data)["forecast"]
```

## 3. Keep package as light as possible

So I don't have to learn a new API and you can maintain as small a package as possible.

- These model-building functions as the core + surgery stuff
- Helpful `numpyro_forecast.extras` which are useful for time series (metrics (CRPS, coverage, pinball etc), backtest, features, datasets, `acf`/`pacf` (although this might be in statsmodels))
- Remove the `fit_svi` / `fit_mcmc` as it isn't forecast specific

I don't think we need `to_datatree`, or it should just be covered where possible by `arviz_base.from_numpyro(posterior=mcmc, posterior_predictive=..., predictions=...)` if that does dims sufficiently well. We don't need `to_datatree` to run everything.

`arviz>=1.2` is currently a hard dependency. I'd make conversion the optional extra (`numpyro_forecast[arviz]`) for a user who wants the model functions + backtest.

## 4. Dims beyond `(time, obs)`

I wondered whether the layout caps you at two dims for the `innovations`/`markov` etc part or if I can try a bigger hierarchy like e.g. forecasting over a grid (business-region combinations) with a plated hierarchy. Reading the hierarchical examples: it doesn't, and `hierarchical_forecasting_2` is exactly this (origin x destination, plates at `dim=-3` and `dim=-1`, `time_series` inside an enclosing plate). Batch dims stack leftward, fine.

One edge worth documenting loudly: `markov_time_series` rejects enclosing plates and takes a `plates=` argument opened inside the scan body instead, so the plate idiom flips between the two latent functions.

## 5. `cast` doesn't feel natural

Lines like

```py
seasonal = cast("Array", origin_seasonal + destin_seasonal)
```

appear in the example notebooks and no human would write them unprompted. Any way to get round this or is `ty` too strict?

## 6. backtest

`backtest_vectorized` fuses all windows into one vmapped SVI run. But using `vmap` is lockstep parallelism (wait-for-the-slowest). This has the same problem as `vmap` vs `pmap` for multiple chains (see [blackjax](https://blackjax-devs.github.io/blackjax/examples/howto_sample_multiple_chains.html)) and means no per-window early stopping inside a fused scan, so a window that would converge in 200 steps pays the full 1,001.

It's embarrassingly parallel, so I would use multiprocessing/joblib, or an orchestrator (I do this sort of thing with Flyte at work).

Separately, `backtest_vectorized` doesn't use `fit_svi`. It builds `SVI(...)` and hand-rolls the `lax.scan` update loop. So there are two SVI code paths in the library (although I would drop anything that isn't in numpyro).

**Comment by juanitorduz** [comment]
Thank you @theorashid ! this was closed by https://github.com/juanitorduz/numpyro_forecast/pull/101 

Happy to get more feedback :)

**Comment by theorashid** [comment]
I'll have a read through. At first glance, I like the notebooks. How do you feel about the changes? anything you like / dislike?

**Comment by juanitorduz** [comment]
I like it's lighter and I do like fit_mcmc is gone as we have make it easy to simply use pure NumPyro 🙂 .

After this bug refactor I wanna keep bringing more examples
